modules/imgcodecs/src/loadsave.cpp

```diff
@@ -1032,6 +1035,222 @@ bool haveImageWriter( const String& filename )
     return !encoder.empty();
 }
 
+class ImageCollection::IImageCollection {
+public:
+    IImageCollection() = default;
+    IImageCollection(const std::string&  filename, int flags);
+    void setup(const String& img, int flags);
+    size_t size() const;
+    Mat read();
+    int width() const;
+    int height() const;
+    bool readHeader();
+    Mat readData();
+    bool advance();
+    int currentIndex() const;
+    void reinitDecoder();
+
+private:
+    String m_filename;
+    int m_flags{};
+    int m_size{};
+    int m_width{};
+    int m_height{};
+    int m_current{};
+    ImageDecoder m_decoder;
+};
+
+ImageCollection::IImageCollection::IImageCollection(std::string const& filename, int flags) {
+    this->setup(filename, flags);
+}
+
+void ImageCollection::IImageCollection::setup(String const& filename, int flags) {
```

Comment:
We need this because if user wants to do std::vector<ImageCollection>. Renaming to `init` might be better